src/syscall_sdl.c

```diff
@@ -276,3 +339,222 @@ void syscall_submit_queue(riscv_t *rv)
+     * from applications
+     */
+    uint32_t sfx_data_offset = *((uint32_t *) &sfxinfo);
+    uint32_t sfx_data_size = *(uint32_t *) ((uint32_t *) &sfxinfo + 1);
+    uint8_t sfx_data[SFX_SAMPLE_SIZE];
+    memory_read(s->mem, sfx_data, sfx_data_offset,
+                sizeof(uint8_t) * sfx_data_size);
+
+    sound_t sfx = {
+        .data = sfx_data,
+        .size = sfx_data_size,
+        .volume = volume,
+    };
+    pthread_create(&sfx_thr, NULL, sfx_handler, &sfx);
+}
+
+static void play_music(riscv_t *rv)
+{
+    state_t *s = rv_userdata(rv); /* access userdata */
+
+    const uint32_t musicinfo_addr = (uint32_t) rv_get_reg(rv, rv_reg_a1);
+    int looping = rv_get_reg(rv, rv_reg_a2);
+    int volume = rv_get_reg(rv, rv_reg_a3);
+
+    musicinfo_t musicinfo;
+    memory_read(s->mem, (uint8_t *) &musicinfo, musicinfo_addr,
+                sizeof(musicinfo_t));
+
+    /*
+     * data and size in application must be placed at
+     * first two fields in structure so that it makes emulator
+     * compatible to any applications when accessing different sfxinfo_t
+     * from applications
+     */
+    uint32_t music_data_offset = *((uint32_t *) &musicinfo);
+    uint32_t music_data_size = *(uint32_t *) ((uint32_t *) &musicinfo + 1);
+    uint8_t music_data[MUSIC_MAX_SIZE];
+    memory_read(s->mem, music_data, music_data_offset, music_data_size);
+
+    sound_t music = {
+        .data = music_data,
+        .size = music_data_size,
+        .looping = looping,
+        .volume = volume,
+    };
+    pthread_create(&music_thr, NULL, music_handler, &music);
+}
+
+static void stop_music(riscv_t *rv UNUSED)
+{
+    if (Mix_PlayingMusic())
+        Mix_HaltMusic();
+}
+
+static void set_music_volume(riscv_t *rv)
+{
+    int volume = rv_get_reg(rv, rv_reg_a1);
+
+    /* multiplied by 8 because volume's max is 15 */
+    Mix_VolumeMusic(volume * 8);
+}
+
+static void init_audio()
+{
+    if (!(SDL_WasInit(-1) & SDL_INIT_AUDIO)) {
+        if (SDL_Init(SDL_INIT_AUDIO) != 0) {
+            fprintf(stderr, "Failed to call SDL_Init()\n");
+            exit(1);
+        }
+    }
+
+    /* sfx samples buffer */
+    sfx_samples = malloc(SFX_SAMPLE_SIZE);
+
+    // Initialize SDL2 Mixer
+    if (Mix_Init(MIX_INIT_MID) != MIX_INIT_MID) {
+        fprintf(stderr, "Mix_Init failed: %s\n", Mix_GetError());
+        exit(1);
+    }
+    if (Mix_OpenAudio(SAMPLE_RATE, AUDIO_U8, CHANNEL_USED, CHUNK_SIZE) == -1) {
+        fprintf(stderr, "Mix_OpenAudio failed: %s\n", Mix_GetError());
+        Mix_Quit();
+        exit(1);
+    }
+}
+
+static void shutdown_audio(riscv_t *rv)
+{
+    stop_music(rv);
+    Mix_HaltChannel(-1);
+    Mix_CloseAudio();
+    Mix_Quit();
+    free(mid_data);
+    free(sfx_samples);
+}
+
+void syscall_setup_audio(riscv_t *rv)
+{
+    /* setup_audio(request) */
+    const int request = rv_get_reg(rv, rv_reg_a0);
+
+    switch (request) {
+    case INIT_AUDIO:
+        init_audio();
+        break;
+    case SHUTDOWN_AUDIO:
```

Comment:
I am unsure whether it is meaningful to offer operations like initialization and shutdown for users, considering that the emulator manages the actual lifecycle of the audio instance. Consider the framebuffer management system call, `draw_frame`: the existence of a single system call is due to the fact that we only enable users to update the available screen, without providing functionality for initializing or closing the window.

Reply:
Document the above considerations in `docs/` directory.